tests/openedx_tagging/core/tagging/test_views.py

```diff
@@ -0,0 +1,299 @@
+            self.assertGreaterEqual(response.data.items(), create_data.items())
+            url = reverse(
+                "oel_tagging:taxonomy-detail", kwargs={"id": response.data["id"]}
+            )
+            response = self.client.get(url)
+            self.assertGreaterEqual(response.data.items(), create_data.items())
+
+    def test_create_taxonomy_class(self):
+        url = reverse("oel_tagging:taxonomy-list")
+
+        create_data = {
+            "name": "taxonomy_data_3",
+            "description": "This is a description",
+            "enabled": False,
+            "required": True,
+            "allow_multiple": True,
+            "taxonomy_class": 'openedx_tagging.core.tagging.models.Taxonomy'
+        }
+
+        self.client.force_authenticate(user=self.staff)
+        response = self.client.post(url, create_data)
+        self.assertEqual(response.status_code, status.HTTP_201_CREATED)
+
+        self.assertGreaterEqual(response.data.items(), create_data.items())
+        url = reverse(
+            "oel_tagging:taxonomy-detail", kwargs={"id": response.data["id"]}
+        )
+        response = self.client.get(url)
+        self.assertGreaterEqual(response.data.items(), create_data.items())
+
+    @ddt.data(
+        {},
+        {"name": "Error taxonomy 2", "required": "Invalid value"},
+        {"name": "Error taxonomy 3", "enabled": "Invalid value"},
+    )
+    def test_create_taxonomy_error(self, create_data):
+        url = reverse("oel_tagging:taxonomy-list")
+
+        self.client.force_authenticate(user=self.staff)
+        response = self.client.post(url, create_data)
+        self.assertEqual(response.status_code, status.HTTP_400_BAD_REQUEST)
+
+    @ddt.data({"name": "System defined taxonomy", "system_defined": True})
+    def test_create_taxonomy_system_defined(self, create_data):
+        """
+        Cannont create a taxonomy with system_defined=true
+        """
+        url = reverse("oel_tagging:taxonomy-list")
+
+        self.client.force_authenticate(user=self.staff)
+        response = self.client.post(url, create_data)
+        self.assertEqual(response.status_code, status.HTTP_201_CREATED)
+        self.assertEqual(response.data["system_defined"], False)
+
+    @ddt.data(
+        (None, status.HTTP_403_FORBIDDEN),
+        ("user", status.HTTP_403_FORBIDDEN),
+        ("staff", status.HTTP_200_OK),
+    )
+    @ddt.unpack
+    def test_update_taxonomy(self, user_attr, expected_status):
+        taxonomy = Taxonomy.objects.create(
+            name="test update taxonomy",
+            description="taxonomy description",
+            enabled=False,
+        )
+        taxonomy.save()
+
+        url = reverse("oel_tagging:taxonomy-detail", kwargs={"id": taxonomy.id})
+        user = getattr(self, user_attr) if user_attr else None
+        if user:
+            self.client.force_authenticate(user=user)
+        response = self.client.put(url, {"name": "new name"})
```

Comment:
Could you also try updating `system_defined` here? It shouldn't change anything (see below):
```suggestion
        response = self.client.put(url, {"name": "new name", "system_defined": True})
```

Reply:
I'm not sure if I understood you here.

You are asking to remove the `test_update_taxonomy_system_defined` and add the proposed line here (in `test_update_taxonomy`?

Reply:
Nope, not exactly.. I wanted to verify that you cannot update the `system_taxonomy` field using the REST API.

So please add/update tests to verify:

* if you create a taxonomy via the API, it will set `system_defined=False`
* if you `post` a change to taxonomy that tries to set  `system_defined=True`, the request should succeed, but the field won't be updated -- which you can see when you re-fetch the taxonomy.
* same as ^ for [`patch`](https://github.com/open-craft/openedx-learning/pull/4#discussion_r1267530631)

Reply:
We already have one of these test:

Create:

https://github.com/open-craft/openedx-learning/blob/963841c1cf9625b0e6bace6f17e3475e9c1642c4/tests/openedx_tagging/core/tagging/test_views.py#L161-L171

We have this update but it is another Test Case:
https://github.com/open-craft/openedx-learning/blob/963841c1cf9625b0e6bace6f17e3475e9c1642c4/tests/openedx_tagging/core/tagging/test_views.py#L203-L212

(It returns 403 because of the rule here:)
https://github.com/open-craft/openedx-learning/blob/963841c1cf9625b0e6bace6f17e3475e9c1642c4/openedx_tagging/core/tagging/rules.py#L25-L32

--

I will create a new test for the condition that you mentioned here for UPDATE and PATCH!

Reply:
@rpenido Yay! We're on the same page now. Your `test_update_taxonomy_system_defined` test is exactly what I needed.

> (It returns 403 because of the rule here:)

👍  yep, that's correct -- the REST API should not allow updating of system-defined taxonomies, and your existing tests are sufficient for testing this.

All good here!